### PR TITLE
FEATURE: Using `@reduxjs/toolkit` in mobile codebase

### LINKED ISSUE #375: Mobile `Filter` component in `Contribute` screen is laggy
**Describe the bug**
When we have many contributions cards (>10), the filter component starts to lag

**To Reproduce**

1. Open dzcode mobile app
2. Navigate to `Contribute` screen
3. Open the filter component
4. See that it lags when trying to open/close it
5. It also lags when you try to select any filter option.

**Expected behavior**
It should behave instant and smooth.

**Environment:**

- OS: iOS 15.0
- Browser : Native Mobile App
- Version: latest version as of writing this issue.

**Additional context**

This is most because, whenever we change the filter options, we trigger a re-render of the whole Contribute screen, including all the contributions cards, you will notice that it's faster if you have fewer contribution cards.

Also, this is even slower on my low-end Android device.

### LINKED ISSUE #480: DX Improvement
**Description**

Improve dev experience a bit, by updating our tooling and the packages we use

**Check List**

- [x] #431
- [x] #482 
- [x] #490
- [x] #486
- [x] #492

### LINKED ISSUE #482: Use Redux Toolkit
**Description**

use Redux toolkit on both `./web` and `./mobile`

**Check List**

1. [ ] #484 
2. [x] #481 

**Additional Comments**

- https://redux-toolkit.js.org/

### PR DESCRIPTION
**Pull request by omdxp**
# Description

This PR is dedicated to use [Redux Toolkit](https://redux-toolkit.js.org) in favor of Redux. It'll help with DX as well as eliminating some current performance issues.

Resolves #482 
Fixes #375 

## Type of change

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by ZibanPirate** [comment]
You really wanted this change haha

I was thinking of doing something like:

https://github.com/ZibanPirate/trove/tree/main/src/redux

Where we have slices in one file, in the same dir as store configuration as well as our homemade subscription function.

And we have action creators in different dir (thunks) where they can be tree shaked and fetched separately from main bundle in the chunk of whatever page required it.

There are some black magic typing there, but it's essentially just helping mapping the slices into one state interface, that can also be used for our own useSliceSelector

I may have gone overboard with my own approach, so I would like to see how you gonna approach this, and have a collaborative approach

**Comment by Omar-Belghaouti** [comment]
> You really wanted this change haha
> 
> I was thinking of doing something like:
> 
> https://github.com/ZibanPirate/trove/tree/main/src/redux
> 
> Where we have slices in one file, in the same dir as store configuration as well as our homemade subscription function.
> 
> And we have action creators in different dir (thunks) where they can be tree shaked and fetched separately from main bundle in the chunk of whatever page required it.
> 
> There are some black magic typing there, but it's essentially just helping mapping the slices into one state interface, that can also be used for our own useSliceSelector
> 
> I may have gone overboard with my own approach, so I would like to see how you gonna approach this, and have a collaborative approach

I will try to make it as similar as the old codebase, however in this new structure the actions and reducers are merged in slices folder, so each slice folder will contains its specific actions. It's a good idea to use RTK Query in this case, but I'll start with using extra reducers as thunk actions and probably later on we'll switch to RTK Query which I think it's the best approach. 

You can find here my approaches in RTK Query:
[https://github.com/Omar-Belghaouti/blogs-full-stack/tree/main/frontend/src/store](https://github.com/Omar-Belghaouti/blogs-full-stack/tree/main/frontend/src/store)
[https://github.com/Omar-Belghaouti/blogs-full-stack/tree/main/frontend/src/store](https://github.com/Omar-Belghaouti/blogs-full-stack/tree/main/frontend/src/store)

As said above, for this PR I'll work on it without RTK Query however it'll have similar structure. And you can work in the web codebase in the meantime I work in the mobile codebase.

**Comment by Omar-Belghaouti** [comment]
@ZibanPirate checkout the new performance for the contribution filters. Now with RTK we can store array data in a normalized form with the help of entity adapters, I was actually planning to normalize it manually [here](https://github.com/dzcode-io/dzcode.io/pull/430#issuecomment-1172960699) 😅 Thanks to RTK it's much easier to normalize data which guarantees performance for arrays data.

**Review by ZibanPirate — COMMENTED** [review]
There are too many tasks in one PR 😅.

This PR is intended to replace Redux with Redux Toolkit, however reviewing the changes, I'm seeing more changes that will lead to many back-and-forths and won't guarantee merging this PR, such as:

- moved `src/redux` to `src/store`, which didn't help the diff, and made it a bit harder to spot the changes.
- refactored the state, and introduced new properties like `status`.
- an attempt to "fix" the contributions filters lag.

I know resisting the urge to refactor the state is hard, but it's better to do this in another PR so it doesn't shadow the intended changes which are just to Replace Redux with Redux Toolkit.

That being said, I will only comment on the Redux to Redux Toolkit changes only:
- Let’s make a clear difference between Actions (that we define in the slice object), and Action Creators (aka Async Actions aka Thunks). meaning that the slice shouldn't know about any Thunk, and updating the `status` should be moved from the slice `.addCase(fetchContributions.pending,` into the Thunk body, and let the Thunk handle the status changes (which again is out of this PR scope to add `status` in the first place). and finally move these thunks outside the state files so they could be tree-shaked (not that needed for mobile code, but better have consistency with web).
- Can we have just one selector? it's not like we have any optimization out of writing a selector for each specific path from the state. It's just too much code I think.

So, let's either address each task at a time in this PR and only merge it once all task changes are reviewed, or have a separate PR for each task, this way is much easier to review and faster to merge.

I have many comments regarding the other changes, but I don't want to add too much noise already to this PR.

i will start working on the web code in the meantime

**Comment by Omar-Belghaouti** [comment]
I know there are many tasks in this PR or should I say it looks like it 😅

> - moved src/redux to src/store, which didn't help the diff, and made it a bit harder to spot the changes.

For this I had to rename the `redux` folder to `store` so there will no ambiguity when importing stuff when using relative imports like we talked about it [here](https://github.com/dzcode-io/dzcode.io/issues/480#issuecomment-1229406329).

> - refactored the state, and introduced new properties like `status`.

The new `status` property is more generalized than the old `refreshing` property. In this version we can definitely know what is happening in the current state (if it's refreshing or not + if there is an error). So you can think of it as it's holding 3 states of the current slice which are `idle` (`refreshing: false`), `loading` (`refreshing: true`) and `error` if there was an error when fetching. The reason it was defined as it is was because of the cases in the `extraReducers` property in the slice, those cases like `pending`, `fulfilled` and `rejected` are defined out of the box for all async thunks which is more suitable in this case.

> - an attempt to "fix" the contributions filters lag.

To be honest I wasn't trying to fix that lag, it was fixed just by implementing the array states as entity adapters that's all. The UI code as you may noticed is not changed at all expect the parts of selecting the data which I thought it's good to have multiple selectors that shows within the code what state are we selecting.

Now, replying to your comments:

> - Let’s make a clear difference between Actions (that we define in the slice object), and Action Creators (aka Async Actions aka Thunks). meaning that the slice shouldn't know about any Thunk, and updating the `status` should be moved from the slice `.addCase(fetchContributions.pending`, into the Thunk body, and let the Thunk handle the status changes (which again is out of this PR scope to add `status` in the first place). and finally move these thunks outside the state files so they could be tree-shaked (not that needed for mobile code, but better have consistency with web).

As said above, those cases are predefined and the `status` property is more suitable for it. I believe there is a state that isn't playing major factor during production which is `error` but I thought it'd definitely be helpful during debugging as it holds the `error.message` if any bug in that situation is occurred. Also, it is recommended to put it this way, I mean to fetch data in the `createAsyncThunk` and only set the data in `extraReducers` otherwise we would have the same old `getState` and `dispatch` which is not ergonomic especially when we have all the capabilities given from `@reduxjs/toolkit`.

> - Can we have just one selector? it's not like we have any optimization out of writing a selector for each specific path from the state. It's just too much code I think.

For this I thought that it is more readable to see which state we are selecting within the code (ex: `useSelector(selectArticles)`). I am not sure if you are saying that we have one selector for the global state or for each one of the slices (that mostly holds the data, status and error). If it's the latter case I will export one selector for each slice so we can have one `useSelector` statement in the UI.

So I hope this clears things up. And let's see your approach on the `web` and do some Discussions afterwards.

**Review by BRAHAIM360 — DISMISSED** [review]
Good job @Omar-Belghaouti
  
I like the idea of using status state.

**Comment by ZibanPirate** [comment]
> > * moved src/redux to src/store, which didn't help the diff, and made it a bit harder to spot the changes.
> 
> For this I had to rename the `redux` folder to `store` so there will no ambiguity when importing stuff when using relative imports like we talked about it [here](https://github.com/dzcode-io/dzcode.io/issues/480#issuecomment-1229406329).

hmm fair, but my only request is to temporarily put the folder name back to `src/redux` in this PR, and we renamed it in a follow-up PR 🙏, I tried adding comments but they seemed out of context because of the diff. Once you rename it, I wil put all my comments for all the tasks in this PR 😄 (and i hope you will have fun addressing them 😈 😅)

**Comment by omdxp** [comment]
Oh I will 😅, I'll rename it back

**Review by ZibanPirate — APPROVED** [review]

File: mobile/src/redux/reducers/contribute-screen/slice.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+interface ContributeScreenState {
+  contributions: EntityState<Model<ContributionEntity, "project">>;
+  filters: EntityState<FilterDto>;
+  status: "idle" | "loading" | "error";
+  error: string | null;
+}
+
+const initialState: ContributeScreenState = {
+  contributions: contributionsAdapter.getInitialState(),
+  filters: filtersAdapter.getInitialState(),
```
[the remark this answers is not in the document]
Comment (omdxp) [reply mobile/src/redux/reducers/contribute-screen/slice.ts:23]: The problem with that approach is that the whole bottom sheet component will completely be rendered when a change occurs which makes some slowness in the contribution screen

**Pull request merged**